plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/validators/N4JSXValidator.xtend

```diff
@@ -355,30 +357,46 @@ class N4JSXValidator extends AbstractN4JSDeclarativeValidator {
 	}
 
 	/**
-	 * Check if there is any attribute in spread operator that is not declared in props
-	 * Notes: this is commented out but not deleted for now since the Stdlib team is still arguing about if this check makes sense
+	 * Check that variable or constant declaration is not named React in N4JSX file to avoid naming clash.
 	 */
-//	def private checkUnknownAttributeInSpreadOperator(JSXSpreadAttribute spreadAttribute, JSXElement jsxElem, Iterable<TMember> attributesInSpreadOperatorType, Iterable<TMember> fieldsOrGettersInProps) {
-//		attributesInSpreadOperatorType.forEach [ attributeInSpreadOperator |
-//			//Look for the field/getter in props that corresponding the spread operator's attribute
-//			val fieldOrGetterInProps = fieldsOrGettersInProps.findFirst[fieldOrGetter | attributeInSpreadOperator.name == fieldOrGetter.name];
-//			if (fieldOrGetterInProps === null) {
-//				val message = IssueCodes.getMessageForJSXSPREADATTRIBUTE_NOT_DECLARED_IN_PROPS(attributeInSpreadOperator.name,
-//					 	jsxElem?.jsxElementName?.expression?.refName
-//					 );
-//						addIssue(
-//							message,
-//							spreadAttribute,
-//							N4JSXPackage.Literals.JSX_SPREAD_ATTRIBUTE__EXPRESSION,
-//							IssueCodes.JSXSPREADATTRIBUTE_NOT_DECLARED_IN_PROPS
-//						);
-//			}
-//
-//		];
-//
-//	}
+	@Check
+	def void checVarDeclNotNamedReact(VariableDeclaration varDecl) {
+		val resourceType = ResourceType.getResourceType(varDecl)
+		// This check is only applicable to N4JSX/JSX file
+		if (!(ResourceType.N4JSX === resourceType || ResourceType.JSX === resourceType))
+			return;
 
+ 		if (varDecl.name == ReactHelper.REACT_NAMESPACE) {
+			val message = IssueCodes.getMessageForJSX_VAR_DECL_NOT_REACT();
+					addIssue(
+						message,
+						varDecl,
+						findNameFeature(varDecl).value,
+						IssueCodes.JSX_VAR_DECL_NOT_REACT
+					);
+		}
+	}
 
+	/**
+	 * Check that function parameter is not named React in N4JSX file to avoid naming clash.
+	 */
+	@Check
+	def void checFormalParameterNotNamedReact(FormalParameter formalParam) {
+		val resourceType = ResourceType.getResourceType(formalParam)
+		// This check is only applicable to N4JSX/JSX file
+		if (!(ResourceType.N4JSX === resourceType || ResourceType.JSX === resourceType))
+			return;
+
+		if (formalParam.name == ReactHelper.REACT_NAMESPACE) {
+			val message = IssueCodes.getMessageForJSX_FORMAL_PARAM_NOT_REACT();
+					addIssue(
+						message,
+						formalParam,
+						findNameFeature(formalParam).value,
+						IssueCodes.JSX_FORMAL_PARAM_NOT_REACT
+					);
+		}
+	}
```

Comment:
These validations do not cover all special cases. For example, the following snippet is still accepted but fails at runtime:
```
class React {}
console.log(<div></div>);
```

Maybe it would be possible to create a single validation for a common super type of VariableDeclaration, FormalParameter, N4ClassDeclaration, etc. (maybe `NamedElement` or `IdentifiableElement`, I'm not sure) and check that for `name!="React"`.